File: guandan_core/lib/src/player.dart

import 'dart:core';

import 'package:guandan_core/guandan_core.dart';

/// Teams in the game
enum PlayerTeam { 
  redTeam, 
  blueTeam;

  static PlayerTeam fromName(String name) {
    return PlayerTeam.values.firstWhere((e) => e.toString().split('.').last == name);
  }
}

/// Positions of the players in the game, relative to a player (anchor player)
enum PlayerPosition { 
  self, // the current player
  leftOpponent, // for 4 and 6 players
  teamMate, // for 4 players only
  rightOpponent, // for 4 and 6 players
  oppositeOpponent, // for 6 players only
  leftTeamMate, // for 6 players only
  rightTeamMate; // for 6 players only

  int get value {
    switch (this) {
      case PlayerPosition.self:
        return 0;
      case PlayerPosition.leftOpponent:
        return 3;
      case PlayerPosition.teamMate:
        return 2;
      case PlayerPosition.rightOpponent:
        return 1;
      case PlayerPosition.oppositeOpponent:
        return 4;
      case PlayerPosition.leftTeamMate:
        return 4;
      case PlayerPosition.rightTeamMate:
        return 2;
    }
  }
 }


/// A class representing a player in the game
/// 
/// A player has a unique id, a seat in the game, a team, a name, a hand of cards, a hand of played cards, and a list of hands left to play.
/// 
class Player {

  /// The id of the player
  final String id;

  /// The seat of the player, assigned by the server (optionally in response to a seat change request), between 1-4 or 1-6
  int seat;

  /// The team of the player
  PlayerTeam team;

  /// The display name of the player, for UI purposes only.
  ///
  /// Falls back to [id] when not set.
  String? displayName;

  /// The bot model if this is an AI player, or null if human.
  ///
  /// Examples: `'basicBot'`, `'strongBot'`, `'providerId-botCode-version'`.
  String? botCode;

  /// The display name, falling back to [id].
  String get name => displayName ?? id;

  /// Whether this player is a human (botCode is null).
  bool get isHumanPlayer => botCode == null;

  /// Whether this player is an AI (botCode is not null).
  bool get isAIPlayer => !isHumanPlayer;

  // The cards on hand, to be dynamically updated as the player plays cards in each round
  PokerCardList? _cardsOnHand;

  /// The cards played by the player, to be dynamically updated as the player plays cards in each round
  PokerCardList playedCards = Hand.emptyHand();

  /// Create a player from a json representation based on the 'is_human' and 'bot_model' fields
  /// If the player is human, it is created as a [HumanPlayer], otherwise, it is created as an AIPlayer
  /// If the bot_model is 'basicBot', it is created as a [BasicBot], otherwise, it is created using the factory function registered with the bot_model
  /// If no factory is registered with the bot_model, it is created as a [BasicBot]
  factory Player.fromJson(Map<String, dynamic> jsonData,) {
    var p = Player(
      Player.readId(jsonData),
      Player.readSeat(jsonData),
      Player.readTeam(jsonData),
      displayName: Player.readDisplayName(jsonData),
      botCode: Player.readBotCode(jsonData),
      cardsOnHand: Player.readCardsOnHand(jsonData),
    );
    p.playedCards = Player.readPlayedCards(jsonData);
    return p;
  }

  Player.deepCopy(Player player, {String? newId, bool withPlayedCards = true, bool withCardsOnHand = true})
      : id = newId ?? player.id,
        seat = player.seat,
        team = player.team,
        displayName = player.displayName,
        botCode = player.botCode,
        _cardsOnHand = withCardsOnHand && player._cardsOnHand != null ? PokerCardList.from(player._cardsOnHand!) : null,
        playedCards = withPlayedCards ? PokerCardList.from(player.playedCards) : PokerCardList.empty();

  /// Create a player from another player
  Player.copy(Player player, {String? newId})
      : id = newId ?? player.id,
        seat = player.seat,
        team = player.team,
        displayName = player.displayName,
        botCode = player.botCode,
        _cardsOnHand = player._cardsOnHand,
        playedCards = player.playedCards;

  /// Create a player with the given id, seat, team, and optional fields.
  ///
  /// [displayName] is for UI display only; defaults to null (falling back to [id] via [name]).
  /// [botCode] identifies bots; null means human player.
  Player(this.id, this.seat, this.team, {this.displayName, this.botCode, PokerCardList? cardsOnHand})
      : _cardsOnHand = cardsOnHand;

  /// The json representation of the player
  ///
  /// If [withCardsOnHand] is true, the cards on hand are included in the json representation.
  /// If [withPlayedCards] is true, the played cards are included in the json representation.
  /// If [withPlayerType] is true, additional player type information (bot model and human status) is included.
  Map<String, dynamic> toJson({bool withCardsOnHand = true, bool withPlayedCards = true, bool withPlayerType = false}) {
    Map<String, dynamic> J = {
      'player_id': id,
      'seat': seat,
      'team': team.toString().split('.').last,
    };
    if (displayName != null) {
      J['display_name'] = displayName;
    }
    if (botCode != null) {
      J['bot_model'] = botCode;
    }
    if (withPlayerType) {
      J['is_human'] = isHumanPlayer;
      if (botCode != null) {
        J['bot_model'] = botCode;
      }
    }
    if (withPlayedCards) {
      J['played_cards'] = playedCards.toString();
    }
    if (withCardsOnHand) {
      if (_cardsOnHand != null) {
        J['cards_on_hand'] = _cardsOnHand!.toString();
      }
    }
    return J;
  }

  static String readId(Map<String, dynamic> jsonData) {
    return jsonData['player_id'] as String;
  }

  static int readSeat(Map<String, dynamic> jsonData) {
    return jsonData['seat'] as int;
  }

  /// Reads the display name from JSON. Falls back to the old `profile.nickname`
  /// for backward compatibility with serialized game snapshots.
  static String? readDisplayName(Map<String, dynamic> jsonData) {
    if (jsonData.containsKey('display_name')) {
      return jsonData['display_name'] as String?;
    }
    // Backward compat: read from old `profile` map.
    if (jsonData.containsKey('profile') && jsonData['profile'] != null) {
      final profile = jsonData['profile'] as Map<String, dynamic>;
      return profile['nickname'] as String?;
    }
    return null;
  }

  /// Reads the bot model from JSON. Falls back to the old `profile.bot_model`
  /// for backward compatibility with serialized game snapshots.
  static String? readBotCode(Map<String, dynamic> jsonData) {
    if (jsonData.containsKey('bot_model')) {
      return jsonData['bot_model'] as String?;
    }
    // Backward compat: read from old `profile` map.
    if (jsonData.containsKey('profile') && jsonData['profile'] != null) {
      final profile = jsonData['profile'] as Map<String, dynamic>;
      if (profile.containsKey('bot_model')) {
        return profile['bot_model'] as String?;
      }
    }
    // Backward compat: explicit is_human field.
    if (jsonData.containsKey('is_human')) {
      return jsonData['is_human'] == false ? 'basicBot' : null;
    }
    return null;
  }

  static PlayerTeam readTeam(Map<String, dynamic> jsonData) {
    return PlayerTeam.values.firstWhere((e) => e.toString().split('.').last == jsonData['team']);
  }

  static PokerCardList? readCardsOnHand(Map<String, dynamic> jsonData) {
    return !jsonData.containsKey('cards_on_hand') || jsonData['cards_on_hand'] == null ? null : PokerCardList.fromString(jsonData['cards_on_hand'] as String);
  }

  static PokerCardList readPlayedCards(Map<String, dynamic> jsonData) {
    return jsonData['played_cards'] == null ? PokerCardList.empty() : PokerCardList.fromString(jsonData['played_cards'] as String);
  }


  /// Check if the player has the given hand or cards, counting duplicates
  bool hasCards(dynamic handOrCards) {
    var cardsToCheck = handOrCards is PokerCardList ? handOrCards.cards : handOrCards as List<PokerCard>;
    return _cardsOnHand!.hasCards(cardsToCheck);
  }

  /// Whether the player has at least one card on hand
  bool get hasAtLeastOneCard {
    return cardCountOnHand > 0;
  }

  /// The number of cards on hand
  int get cardCountOnHand {
    if (_cardsOnHand != null) {
      return _cardsOnHand!.length;
    }
    return 27 - playedCards.length;
  }

  /// Play the given hand. This will remove the cards from the player's hand and add them to the played cards.
  void play(Hand hand) {
    if(_cardsOnHand !=null && _cardsOnHand!.isNotEmpty)
    {
      _cardsOnHand!.removeCards(hand.cards);
    }
    playedCards.addAll(hand.cards);
  }

  /// Reset the player's hand and played cards
  void resetHands() {
    _cardsOnHand = null;
    playedCards.clear();
  }

  /// set the player's hand
  void setCardsOnHand(PokerCardList hand) {
    if (hand is Hand) {
      _cardsOnHand = hand;
      return;
    } 
    _cardsOnHand = Hand(hand.cards, HandType.unknown);
  }

  /// Get the cards on hand. Returns null if it is unknown (e.g., for other players in the client side)
  PokerCardList? get cardsOnHand {
    return _cardsOnHand;
  }

  Player deepCopyWith({
    String? id,
  }) {
    return Player.deepCopy(
      this,
      newId: id ?? this.id,
    );
  }
}


/// Get the seat of the player next to the given [seat] in the game, with the given [maxPlayers] number of players.
/// 
/// The seat numbers are 1-based.
int nextSeat(int seat, int maxPlayers) {
  seat = (seat + 1) % maxPlayers;
  if (seat == 0) {
    seat = maxPlayers;
  }
  return seat;
}

/// Get the player next (counter clockwise) to the [currentPlayer] in the game, according to the given criteria. 
/// This never returns the current player.
/// Returns null if no player is found.
Player? nextPlayer(Player currentPlayer, List<Player> players, bool teamMateOnly, bool nonemptyHandOnly) {
  for (int i = 0; i < players.length-1; i++) {
    var seat = nextSeat(currentPlayer.seat + i, players.length);
    var player = players.firstWhere((p) => p.seat == seat);
    if (teamMateOnly && player.team != currentPlayer.team) {
      continue;
    }
    if (nonemptyHandOnly && !player.hasAtLeastOneCard) {
      continue;
    }
    return player;
  }
  return null;
}


/// Get the relative position of a [player] next (counter clockwise) to the [self] in the game.
/// 
/// The [numberOfPlayers] is the total number of players in the game.
PlayerPosition getPlayerPosition(Player self, Player player, int numberOfPlayers) {
  int relativePosition = (player.seat - self.seat + numberOfPlayers) % numberOfPlayers;

  switch (numberOfPlayers) {
    case 4:
      switch (relativePosition) {
        case 0:
          return PlayerPosition.self;
        case 1:
          return PlayerPosition.rightOpponent;
        case 2:
          return PlayerPosition.teamMate;
        case 3:
          return PlayerPosition.leftOpponent;
      }
      break;
    case 6:
      switch (relativePosition) {
        case 0:
          return PlayerPosition.self;
        case 1:
          return PlayerPosition.rightOpponent;
        case 2:
          return PlayerPosition.rightTeamMate;
        case 3:
          return PlayerPosition.oppositeOpponent;
        case 4:
          return PlayerPosition.leftTeamMate;
        case 5:
          return PlayerPosition.leftOpponent;
      }
      break;
  }
  throw ArgumentError('Invalid number of players or player positions');
}

File: py_guandan/guandan_core/player.py
"""Player models compatible with Dart ``player.dart``."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

from .cards import Hand, HandType, PokerCardList


class PlayerTeam(Enum):
    """Teams in a Guandan game."""

    RED_TEAM = "redTeam"
    BLUE_TEAM = "blueTeam"

    @classmethod
    def from_name(cls, name: str) -> "PlayerTeam":
        return cls(name)


class PlayerPosition(Enum):
    """Position relative to an anchor player."""

    SELF = ("self", 0)
    LEFT_OPPONENT = ("leftOpponent", 3)
    TEAM_MATE = ("teamMate", 2)
    RIGHT_OPPONENT = ("rightOpponent", 1)
    OPPOSITE_OPPONENT = ("oppositeOpponent", 4)
    LEFT_TEAM_MATE = ("leftTeamMate", 4)
    RIGHT_TEAM_MATE = ("rightTeamMate", 2)

    def __init__(self, wire_name: str, value_index: int) -> None:
        self.wire_name = wire_name
        self.value_index = value_index


@dataclass
class Player:
    """A player with identity, seat, team, and optional known hand state."""

    id: str
    seat: int
    team: PlayerTeam
    display_name: Optional[str] = None
    bot_code: Optional[str] = None
    cards_on_hand: Optional[PokerCardList] = None
    played_cards: PokerCardList = field(default_factory=PokerCardList.empty)

    @property
    def name(self) -> str:
        return self.display_name or self.id

    @property
    def is_human_player(self) -> bool:
        return self.bot_code is None

    @property
    def is_ai_player(self) -> bool:
        return not self.is_human_player

    @property
    def has_at_least_one_card(self) -> bool:
        return self.card_count_on_hand > 0

    @property
    def card_count_on_hand(self) -> int:
        if self.cards_on_hand is not None:
            return len(self.cards_on_hand)
        return 27 - len(self.played_cards)

    def to_json(
        self,
        with_cards_on_hand: bool = True,
        with_played_cards: bool = True,
        with_player_type: bool = False,
    ) -> dict[str, Any]:
        """Serialize using Dart-compatible JSON keys."""
        data: dict[str, Any] = {
            "player_id": self.id,
            "seat": self.seat,
            "team": self.team.value,
        }
        if self.display_name is not None:
            data["display_name"] = self.display_name
        if self.bot_code is not None:
            data["bot_model"] = self.bot_code
        if with_player_type:
            data["is_human"] = self.is_human_player
            if self.bot_code is not None:
                data["bot_model"] = self.bot_code
        if with_played_cards:
            data["played_cards"] = str(self.played_cards)
        if with_cards_on_hand and self.cards_on_hand is not None:
            data["cards_on_hand"] = str(self.cards_on_hand)
        return data

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> "Player":
        """Create a player from Dart-compatible JSON."""
        player = cls(
            id=cls.read_id(data),
            seat=cls.read_seat(data),
            team=cls.read_team(data),
            display_name=cls.read_display_name(data),
            bot_code=cls.read_bot_code(data),
            cards_on_hand=cls.read_cards_on_hand(data),
        )
        player.played_cards = cls.read_played_cards(data)
        return player

    @staticmethod
    def read_id(data: dict[str, Any]) -> str:
        return data["player_id"]

    @staticmethod
    def read_seat(data: dict[str, Any]) -> int:
        return data["seat"]

    @staticmethod
    def read_display_name(data: dict[str, Any]) -> Optional[str]:
        if "display_name" in data:
            return data.get("display_name")
        profile = data.get("profile")
        return profile.get("nickname") if isinstance(profile, dict) else None

    @staticmethod
    def read_bot_code(data: dict[str, Any]) -> Optional[str]:
        if "bot_model" in data:
            return data.get("bot_model")
        profile = data.get("profile")
        if isinstance(profile, dict) and "bot_model" in profile:
            return profile.get("bot_model")
        if "is_human" in data:
            return "basicBot" if data.get("is_human") is False else None
        return None

    @staticmethod
    def read_team(data: dict[str, Any]) -> PlayerTeam:
        return PlayerTeam.from_name(data["team"])

    @staticmethod
    def read_cards_on_hand(data: dict[str, Any]) -> Optional[PokerCardList]:
        cards = data.get("cards_on_hand")
        return None if cards is None else PokerCardList.from_string(cards)

    @staticmethod
    def read_played_cards(data: dict[str, Any]) -> PokerCardList:
        cards = data.get("played_cards")
        return PokerCardList.empty() if cards is None else PokerCardList.from_string(cards)

    def has_cards(self, hand_or_cards: Any) -> bool:
        """Return whether the known hand contains all requested cards."""
        if self.cards_on_hand is None:
            return False
        cards = hand_or_cards.cards if isinstance(hand_or_cards, PokerCardList) else hand_or_cards
        return self.cards_on_hand.has_cards(cards)

    def play(self, hand: Hand) -> None:
        """Remove played cards from hand and add them to played cards."""
        if self.cards_on_hand is not None and len(self.cards_on_hand) > 0:
            self.cards_on_hand.remove_cards(hand.cards)
        self.played_cards.add_all(hand.cards)

    def reset_hands(self) -> None:
        """Forget current hand and clear played cards."""
        self.cards_on_hand = PokerCardList.empty()
        self.played_cards.clear()

    def set_cards_on_hand(self, hand: PokerCardList) -> None:
        """Set the player's hand."""
        self.cards_on_hand = hand if isinstance(hand, Hand) else Hand(hand.cards, HandType.UNKNOWN)

    def deep_copy_with(self, id: Optional[str] = None) -> "Player":
        """Deep-copy the player, optionally replacing the id."""
        return self.deep_copy(self, new_id=id or self.id)

    @classmethod
    def deep_copy(
        cls,
        player: "Player",
        new_id: Optional[str] = None,
        with_played_cards: bool = True,
        with_cards_on_hand: bool = True,
    ) -> "Player":
        """Deep-copy a player, matching Dart ``Player.deepCopy``."""
        return cls(
            id=new_id or player.id,
            seat=player.seat,
            team=player.team,
            display_name=player.display_name,
            bot_code=player.bot_code,
            cards_on_hand=PokerCardList.from_list(player.cards_on_hand.cards)
            if with_cards_on_hand and player.cards_on_hand
            else None,
            played_cards=PokerCardList.from_list(player.played_cards.cards) if with_played_cards else PokerCardList.empty(),
        )

    @classmethod
    def copy(cls, player: "Player", new_id: Optional[str] = None) -> "Player":
        """Shallow-copy a player, matching Dart ``Player.copy``."""
        return cls(
            id=new_id or player.id,
            seat=player.seat,
            team=player.team,
            display_name=player.display_name,
            bot_code=player.bot_code,
            cards_on_hand=player.cards_on_hand,
            played_cards=player.played_cards,
        )


def next_seat(seat: int, max_players: int) -> int:
    """Return the next 1-based seat number."""
    seat = (seat + 1) % max_players
    return max_players if seat == 0 else seat


def next_player(
    current_player: Player,
    players: list[Player],
    team_mate_only: bool,
    nonempty_hand_only: bool,
) -> Optional[Player]:
    """Return the next player matching the requested criteria."""
    for i in range(len(players) - 1):
        seat = next_seat(current_player.seat + i, len(players))
        player = next(p for p in players if p.seat == seat)
        if team_mate_only and player.team != current_player.team:
            continue
        if nonempty_hand_only and not player.has_at_least_one_card:
            continue
        return player
    return None


def get_player_position(anchor_player: Player, player: Player, number_of_players: int) -> PlayerPosition:
    """Return ``player``'s relative position from ``anchor_player``."""
    relative_position = (player.seat - anchor_player.seat + number_of_players) % number_of_players
    if number_of_players == 4:
        return {
            0: PlayerPosition.SELF,
            1: PlayerPosition.RIGHT_OPPONENT,
            2: PlayerPosition.TEAM_MATE,
            3: PlayerPosition.LEFT_OPPONENT,
        }[relative_position]
    if number_of_players == 6:
        return {
            0: PlayerPosition.SELF,
            1: PlayerPosition.RIGHT_OPPONENT,
            2: PlayerPosition.RIGHT_TEAM_MATE,
            3: PlayerPosition.OPPOSITE_OPPONENT,
            4: PlayerPosition.LEFT_TEAM_MATE,
            5: PlayerPosition.LEFT_OPPONENT,
        }[relative_position]
    raise ValueError("Invalid number of players or player positions")
